src/App/VPSScan/Scan/Core.hs

```diff
@@ -0,0 +1,95 @@
+module App.VPSScan.Scan.Core
+  ( coreAuthHeader
+  , createCoreProject
+  , completeCoreProject
+  , getSherlockInfo
+  , createLocator
+  , createRevisionLocator
+  , buildRevision
+  , SherlockInfo(..)
+  )
+where
+
+import App.VPSScan.Types
+import App.Util (parseUri)
+import Data.Text (pack, Text)
+import Prelude
+import Network.HTTP.Req
+import Data.Text.Encoding (encodeUtf8)
+import Control.Monad.IO.Class (liftIO, MonadIO)
+import Control.Effect.Diagnostics
+import Data.Aeson
+import Data.Time.Clock.POSIX (getPOSIXTime)
+
+data SherlockInfo = SherlockInfo
+  { sherlockUrl :: Text
+  , sherlockClientToken :: Text
+  , sherlockClientId :: Text
+  , sherlockOrgId :: Text
+  }
+  deriving (Eq, Ord, Show)
+
+instance FromJSON SherlockInfo where
+  parseJSON = withObject "" $ \obj -> do
+    auth <- obj .: "auth"
+    sherlockUrl <- obj .: "url"
+    sherlockOrgId <- obj .: "orgId"
+    sherlockClientId <- auth .: "clientId"
+    sherlockClientToken <- auth .: "clientToken"
+    return (SherlockInfo { sherlockUrl = sherlockUrl, sherlockClientId = sherlockClientId, sherlockClientToken = sherlockClientToken, sherlockOrgId = sherlockOrgId })
```

Comment:
This is good! And we can make it a little better:

As we know, `Functor` allows lifting a function `a -> b` into `f a -> f b`:
```hs
class Functor f where
  fmap :: (a -> b) -> (f a -> f b)
```

which we use elsewhere to "lift" newtype wrappers (like `FilterExpressions`) when we parse command line options.

there's an abstraction, `Applicative` that generalizes this to lifting `N`-ary functions:
```hs
-- this is the same as fmap
liftA1 :: Applicative f => (a -> b) -> (f a -> f b)
liftA1 f a = f <$> a

liftA2 :: Applicative f => (a -> b -> c) -> (f a -> f b -> f c)
liftA2 f a b = f <$> a <*> b

liftA3 :: Applicative f => (a -> b -> c -> d) -> (f a -> f b -> f c -> f d)
liftA3 f a b c = f <$> a <*> b <*> c

liftA4 :: Applicative f => (a -> b -> c -> d -> e) -> (f a -> f b -> f c -> f d -> f e)
liftA4 f a b c d = f <$> a <*> b <*> c <*> d
```

 (the right-side parentheses are redundant, but that's not really important)

we notice a pattern here: for each new argument, we chain another `<*> val`

---

how does this apply here? well, it turns out that the `SherlockInfo` data constructor is a function:
```hs
SherlockInfo :: Text -> Text -> Text -> Text -> SherlockInfo
```

and we have a bunch of aeson `Parser`s for the fields:
```hs
obj .: "url" :: Parser Text
obj .: "orgId" :: Parser Text
...
```

and we're building a `Parser SherlockInfo`. squinting at `liftA4` again:
```hs
liftA4 :: (Text -> Text -> Text -> Text -> SherlockInfo)
       -> Parser Text
       -> Parser Text
       -> Parser Text
       -> Parser Text
       -> Parser SherlockInfo
```

or applying it directly:
```hs
liftA4 SherlockInfo (obj .: "url") (obj .: "orgId") (auth .: "clientId") (auth .: "clientToken")
```

and inlining the definition of `liftA4`:
```hs
SherlockInfo <$> obj .: "url"
             <*> obj .: "orgId"
             <*> auth .: "clientId"
             <*> auth .: "clientToken"
```

---

bringing it all together:
```hs
instance FromJSON SherlockInfo where
  parseJSON = withObject "SherlockInfo" $ \obj -> do
    auth <- obj .: "auth"

    SherlockInfo
      <$> obj .: "url"
      <*> obj .: "orgId"
      <*> auth .: "clientId"
      <*> auth .: "clientToken"
```

---

it may not be immediately clear why this refactor matters. the primary motivator is that using `Applicative` gives us _less_ power, so there's substantially less mental overhead in determining what this does, whether it's correct, and making modifications to it later. we can discuss this later if you'd like, with examples